lib/api/rule.go

```diff
@@ -23,34 +23,109 @@ import (
 	"gopkg.in/go-playground/validator.v8"
 )
 
+// A Rule encapsulates a set of match criteria and an action.  Both tiered security Policy
+// and security Profiles reference rules - separated out as a list of rules for both
+// ingress and egress packet matching.
+//
+// Each positive match criteria has a negated version, prefixed with ”Not”. All the match
+// criteria within a rule must be satisfied for a packet to match. A single rule can contain
+// the positive and negative version of a match and both must be satisfied for the rule to match.
 type Rule struct {
 	Action string `json:"action" validate:"action"`
 
+	// Protocol is an optional field that restricts the rule to only apply to traffic of
+	// a specific IP protocol. Required if any of the EntityRules contain Ports
+	// (because ports only apply to certain protocols).
+	//
+	// Must be one of these string values: "tcp", "udp", "icmp", "icmpv6", "sctp", "udplite"
+	// or an integer in the range 1-255.
 	Protocol *Protocol   `json:"protocol,omitempty" validate:"omitempty"`
+
+	// ICMP is an optional field that restricts the rule to apply to a specific type and
+	// code of ICMP traffic.  This should only be specified if the Protocol field is set to
+	// "icmp" or "icmpv6".
 	ICMP     *ICMPFields `json:"icmp,omitempty" validate:"omitempty"`
 
+	// NotProtocol is the negated version of the Protocol field.
 	NotProtocol *Protocol   `json:"!protocol,omitempty" validate:"omitempty"`
+
+	// NotICMP is the negated version of the ICMP field.
 	NotICMP     *ICMPFields `json:"!icmp,omitempty" validate:"omitempty"`
 
+	// Source contains the match criteria that apply to source entity.
 	Source      EntityRule `json:"source,omitempty" validate:"omitempty"`
+
+	// Destination contains the match criteria that apply to destination entity.
 	Destination EntityRule `json:"destination,omitempty" validate:"omitempty"`
 }
 
 // ICMPFields defines structure for ICMP and NotICMP sub-struct for ICMP code and type
 type ICMPFields struct {
+	// Match on a specific ICMP type.  For example a value of 8 refers to ICMP Echo Request
+	// (i.e. pings).
 	Type *int `json:"type,omitempty" validate:"omitempty,gte=0,lte=255"`
```

Comment:
IIRC, 0 and 255 have special meaning in the kernel, might be best to forbid those.